aineko/datasets/kafka.py

```diff
@@ -0,0 +1,823 @@
+# Copyright 2023 Aineko Authors
+# SPDX-License-Identifier: Apache-2.0
+"""Kafka Dataset.
+
+Contains Kafka dataset, a subclass of AbstractDataset.
+
+The storage layer for a Kafka dataset is a Kafka topic.
+
+The query layer for reading and writing to the topic
+is a Kafka consumer and producer, respectively.
+"""
+import datetime
+import json
+import logging
+import os
+from typing import Any, Dict, List, Literal, Optional, Union
+
+from confluent_kafka import (  # type: ignore
+    OFFSET_INVALID,
+    Consumer,
+    KafkaError,
+    Message,
+    Producer,
+)
+from confluent_kafka.admin import AdminClient, NewTopic  # type: ignore
+from pydantic import BaseModel
+
+from aineko.config import AINEKO_CONFIG, DEFAULT_KAFKA_CONFIG
+from aineko.core.dataset import (
+    AbstractDataset,
+    DatasetCreateStatus,
+    DatasetError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaDatasetError(DatasetError):
+    """General Exception for KafkaDataset errors."""
+
+    pass
+
+
+class KafkaCredentials(BaseModel):
+    """Kafka credentials model."""
+
+    bootstrap_servers: str = DEFAULT_KAFKA_CONFIG.get("BROKER_CONFIG").get(
+        "bootstrap.servers"
+    )
+    security_protocol: Optional[str] = None
+    sasl_mechanism: Optional[str] = None
+    sasl_plain_username: Optional[str] = None
+    sasl_plain_password: Optional[str] = None
+
+
+class ConsumerParams(BaseModel):
+    """Parameters for initializing a Kafka Consumer.
+
+    Passed in as connection_params when calling
+        ```python
+        self.initialize(create_consumer=True,
+                      connection_params=ConsumerParams(...))
+        ```
+    """
+
+    dataset_name: str
+    node_name: str
+    pipeline_name: str
+    prefix: Optional[str] = None
+    has_pipeline_prefix: bool
+    consumer_config: Dict[str, Any] = DEFAULT_KAFKA_CONFIG.get(
+        "CONSUMER_CONFIG"
+    )
+
+
+class ProducerParams(BaseModel):
+    """Parameters for initializing a Kafka Producer.
+
+    Passed in as conection_params when calling
+        ```python
+        self.initialize(create_producer=True,
+                      connection_params=ProducerParams(...))
+        ```
+    """
+
+    dataset_name: str
+    node_name: str
+    pipeline_name: str
+    prefix: Optional[str] = None
+    has_pipeline_prefix: bool
+    producer_config: Dict[str, Any] = DEFAULT_KAFKA_CONFIG.get(
+        "PRODUCER_CONFIG"
+    )
```

Comment:
Can create a generic KafkaParams class which both of these inherit so we don't need to repeat the params

Reply:
That seems cleaner, thanks.